crates/solvers/src/domain/solver/baseline.rs

```diff
@@ -69,80 +69,109 @@ impl Baseline {
         // not lock up the [`tokio`] runtime and cause it to slow down handling
         // the real async things. For larger settlements, this can block in the
         // 100s of ms.
+        tracing::trace!("solving {} orders", auction.orders.len());
+
+        let (sender, mut receiver) = tokio::sync::mpsc::channel(auction.orders.len());
```

Comment:
The over all strategy makes sense to me but I think the code can probably be collapsed a lot.
Most of the time we don't get a solution for a lot of orders. Therefore I would go for `mpsc::unbounded_channel()`. (BTW since you continuously queried the `receiver` a `oneshot` channel would probably have been fine as well)
When we have that we can probably make things more concise by doing something like:
```rust
        let (sender, mut receiver) = tokio::sync::mpsc::unbounded_channel();
        let time_left = tokio::time::Duration::from_millis(150); // (actually compute from auction deadline)
        if tokio::time::timeout(time_left, self.0.solve(auction, sender)).await.is_err() {
            tracing::debug!("reached timeout while solving orders");
        }
        let mut solutions = vec![];
        while let Ok(solution) = receiver.try_recv() {
            solutions.push(solution);
        }
        tracing::trace!("solved {} orders", solutions.len());
```
Technically this will probably run into issues if we encounter an endless loop in the worker thread. If we are worried about that I think we should continue to use `tokio::task::spawn_blocking` and if not doing everything in the same task seems fine to me.

Reply:
i think this doesn't fix the `500ms` issue I mentioned in the description ☝️ 
`tokio::time::timeout` will not return at exactly `time_left` but just after each order is finished.
That's why I introduced the outer loop in main thread to control the maximum time I want to spend in background thread.

Reply:
Yeah, this is why I added the caveat about the endless loop (or your 500ms example). But the main point was: Either we are worried about it then we should continue to use `spawn_blocking` or we are not worried about it then we can probably get away with not spawning anything and just have a few ms time buffer.
But the the compressed code would still work with `task::spawn_blocking()`:
```rust
        let inner = self.0.clone();
        let task = || async move { inner.solve(auction, sender).await };
        let timeout_future = tokio::time::timeout(time_left, tokio::task::spawn_blocking(task));
        if timeout_future.await.is_err() {
        // ...
```
This should be equivalent to your handrolled loop since `poll`ing the spawned task does not block if something inside the spawned task is blocking so the `timeout` should be spot on.

Reply:
Can't make it work: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=992e9102decd4d87eab797dc00800b31

Reply:
Curious that `spawn_blocking` indeed does not seem to work but using `spawn` does. ([example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=06d555a08901a4ef128adfe2a0a7edba))
AFAIK these function should behave the same with the difference that `spawn_blocking` takes a thread from a thread pool where long blocking is not supposed to be a big issue.

Reply:
Yeah. I refactored the solution to be aligned with the example. Also, `unbounded_channel` makes sense.